Sources/TuistLoader/Loaders/RecursiveManifestLoader.swift

```diff
@@ -65,8 +65,13 @@ public class RecursiveManifestLoader: RecursiveManifestLoading {
         var paths = Set(paths)
         while !paths.isEmpty {
             paths.subtract(cache.keys)
-            let projects = try Array(paths).map(context: ExecutionContext.concurrent) {
-                try manifestLoader.loadProject(at: $0)
+            let projects = try Array(paths).map(context: ExecutionContext.concurrent) { path -> ProjectDescription.Project in
+                let serializedProject = path.appending(component: "Project.json")
+                if fileHandler.exists(serializedProject) {
+                    return try JSONDecoder().decode(ProjectDescription.Project.self, from: try fileHandler.readFile(serializedProject))
+                } else {
+                    return try manifestLoader.loadProject(at: path)
+                }
```

Comment:
I feel like this is complicates this component 😞 _(bear with me I haven't been following dependencies work closely so apologies if this has already been discussed at length)_

There's a few considerations to keep in mind:

- Serializing manifests is something that is already taken care off via `CachedManifestLoader`, that component also includes hashes and versioning information in the cached json manifest to ensure they are invalided when appropriate. This sadly won't benefit from that.
- This could accidentally allow user defined `Project.json` locally 🤐 which would be prioritised over the `Project.swift`
- The recursive manifest loader isn't the only manifest loader in use today, there's a few different components that are used in different contexts
    - `RecursiveManifestLoader`: Loads all manifests given a starting manifest path(s)
    - `CachedManifestLoader`: Deals with caching manifests to minimise the times where we perform a swift compilation on the Swift manifests 
    - `ManifestLoader`: The main single Swift manifest loader  (performs the compilation needed)

Given this, indeed `RecursiveManifestLoader` _could be_ a place to wire in the dependencies logic. Alternatively it can be kept as a separate / independent step in the pipeline allowing those component to be single purpose. A [suggestion I had _(not sure on feasibility)_ ](https://github.com/tuist/tuist/pull/3093/files#r656816395) is to to reuse the `RecurisiveManifestLoader` in a secondary pass once we've collected all the `.external > .project()` dependency paths, and load all those manifests independently. Once loaded we can add those additional manifests to the collection of all manifests we have loaded so far before proceeding with the subsequent generation pipeline steps.

Having it independent could also allow writing a dedicated external manifest loader purely for source based dependencies if needed _(e.g. if they are always json based or may have additional metadata)_.

This my current understanding of the workflow:

![Dependencies Graph](https://user-images.githubusercontent.com/11914919/123223262-a7481980-d4c8-11eb-82db-93e5ce251149.png)

Reply:
Thanks for your comment!

I have moved the logic to load dependencies project outside of the `RecursiveManifestLoader`, and also moved the logic to load serialized project to the `ManifestLoader` (and handled it also for the `CachedManifestLoader`.

What do you think of the new solution? 🙌